python/dgl/graph_index.py

```diff
@@ -1,43 +1,44 @@
 """Module for graph index class definition."""
 from __future__ import absolute_import
 
-import ctypes
 import numpy as np
 import networkx as nx
 import scipy
 
-from ._ffi.base import c_array
+from ._ffi.object import register_object, ObjectBase
 from ._ffi.function import _init_api
 from .base import DGLError
 from . import backend as F
 from . import utils
 
-GraphIndexHandle = ctypes.c_void_p
-
 class BoolFlag(object):
     """Bool flag with unknown value"""
     BOOL_UNKNOWN = -1
     BOOL_FALSE = 0
     BOOL_TRUE = 1
 
-class GraphIndex(object):
+@register_object('graph.Graph')
+class GraphIndex(ObjectBase):
     """Graph index object.
 
-    Parameters
-    ----------
-    handle : GraphIndexHandle
-        Handler
+    Note
+    ----
+    Do not create GraphIndex directly, you can create graph index object using
+    following functions:
+
+    - `dgl.graph_index.from_edge_list`
+    - `dgl.graph_index.from_scipy_sparse_matrix`
+    - `dgl.graph_index.from_networkx`
+    - `dgl.graph_index.from_shared_mem_csr_matrix`
+    - `dgl.graph_index.from_csr`
+    - `dgl.graph_index.from_coo`
     """
-    def __init__(self, handle):
-        self._handle = handle
-        self._multigraph = None  # python-side cache of the flag
-        self._readonly = None  # python-side cache of the flag
-        self._cache = {}
-
-    def __del__(self):
-        """Free this graph index object."""
-        if hasattr(self, '_handle'):
-            _CAPI_DGLGraphFree(self._handle)
+    def __new__(cls):
+        obj = ObjectBase.__new__(cls)
+        obj._multigraph = None  # python-side cache of the flag
+        obj._readonly = None  # python-side cache of the flag
+        obj._cache = {}
+        return obj
```

Comment:
this method is called in the graph index CAPI?

Reply:
What's these flag for?

Reply:
Nop. This is called on python side when a CAPI returns an Object. See here: https://github.com/dmlc/dgl/blob/master/python/dgl/_ffi/_ctypes/object.py#L20

The flags indicate whether the graph is a multigraph or a readonly graph. We could use CAPI to get the flag. Here it also saves a copy (a bool value) in python.